### PR TITLE
feat: Add object detection

### PR DESCRIPTION
**Pull request by intsco**
Model API for object detection. Extracting code shared between object detection and classification.

* Refactor API composition code to separate type of model serving from ML model type.
* Separate `DetectionObject` and `SegmentationObject` classes for object detection and instance segmentation.
* Unify input of object detection and classification models.
* Model API documentation for classification.
* Update classification model API output format.
* API tests for object detection and classification models.

### CODE REVIEW

File: telesto/classification/app.py [hunk 1]
```diff
@@ -1,76 +1,80 @@
...
+    "name": "Predict endpoint",
+    "request_body": {
+        "images": [
+            {
+                "content": "<str>"
+            }
+        ],
+    },
+    "image_content_format": {
+        **PNG_RGB_GRAY8_BASE64_FORMAT,
```
Comment (koosk) [inline telesto/classification/app.py:31]: Can we extend this later to more formats? 16 bit tiff is very common in microscopy images.

File: telesto/instance_segmentation/__init__.py [hunk 2]
```diff
@@ -63,11 +51,21 @@ def __init__(self, coords: List[Tuple[int, int]]):
...
-    def __eq__(self, other: "DetectionObject"):
+    def __eq__(self, other: "SegmentationObject"):
         return other.coords == self.coords and other.bbox == self.bbox
 
+    def asdict(self, image_size: Tuple[int, int]) -> Dict:
+        dic = {
+            "x": self.bbox.x1,
+            "y": self.bbox.y1,
+            "w": self.bbox.x2 - self.bbox.x1 + 1,
+            "h": self.bbox.y2 - self.bbox.y1 + 1,
```
Comment (koosk) [inline telesto/instance_segmentation/__init__.py:64]: Are you sure these +1s are needed? X2 - X1 should give the width, regardless of 0-based indexing.

File: telesto/object_detection/app.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+                    {
+                        "x": "<int>",
+                        "y": "<int>",
+                        "w": "<int>",
+                        "h": "<int>",
+                    }
+                ]
+            }
+        ]
+    },
```
Comment (koosk) [inline telesto/object_detection/app.py:46]: Is the `class` missing here? The json has a `classes` field right below, so I guess every object should have this field as well.

File: telesto/object_detection/app.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    "classes": config.get("common", "classes").split(","),
+}
+
+
+def preprocess(doc: Dict) -> List[np.ndarray]:
+    input_list = convert_base64_images_to_arrays(doc)
+    if not (0 < len(input_list) <= 32):
```
Comment (koosk) [inline telesto/object_detection/app.py]: Is there a reason behind the limit of 32? Could it be a parameter later?

**Review by koosk — APPROVED** [review]
When multiple images are passed in a request, would be great if some custom identifier could be associated with them. Right now, I assume, the response is paired with the input based on their order only. Some text field could be added, which the user could use to add filenames, hashes, or whatever.

I have some concerns that the object detection response does not return the predicted class. I have added some comments on this, pls check.

File: telesto/classification/app.py (hunk 1 above)
Comment (intsco) [reply telesto/classification/app.py:31]: Yes, we can definitely extend the list of supported formats. Pillow actually already supports 16-bit TIFF images https://pillow.readthedocs.io/en/5.1.x/releasenotes/4.3.0.html#loading-16-bit-tiff-images

File: telesto/instance_segmentation/__init__.py (hunk 2 above)
Comment (intsco) [reply telesto/instance_segmentation/__init__.py:64]: Discussed during the last meeting.

File: telesto/object_detection/app.py (hunk 3 above)
Comment (intsco) [reply telesto/object_detection/app.py:46]: Yes, we will need support of classes. However, it'll be added separately.

**Comment by intsco** [comment]
>When multiple images are passed in a request, would be great if some custom identifier could be associated with them. Right now, I assume, the response is paired with the input based on their order only. Some text field could be added, which the user could use to add filenames, hashes, or whatever.

@koosk While I support the idea of making sure that the predictions match the inputs, I am not sure whether we should force clients to provide images ids. It feels like asking a bit more than needed. We could ensure it internally instead.

Or did you have any other reasons in mind why we should accept image ids?

**Comment by koosk** [comment]
> I am not sure whether we should force clients to provide images ids.

We definitely should not force them. It would be just an option: if they provide anything as an id with the images, we return the same id in the response. It would make matching much easier. Let's say they have a very special filename format (like in microscopy) and want to compare the inputs with the result images manually, they could save the response image with the given id. Otherwise, they might have to check the ordering which is harder, and they might want to rename it anyway. This, when combined with polling makes it quite uncomfortable to use.

**Comment by intsco** [comment]
> > I am not sure whether we should force clients to provide images ids.
> 
> We definitely should not force them. It would be just an option: if they provide anything as an id with the images, we return the same id in the response. It would make matching much easier. Let's say they have a very special filename format (like in microscopy) and want to compare the inputs with the result images manually, they could save the response image with the given id. Otherwise, they might have to check the ordering which is harder, and they might want to rename it anyway. This, when combined with polling makes it quite uncomfortable to use.

Ok, it's a valid point. I'd prefer to add it in a separate PR though. I've created an issue for this.

**Pull request merged**